icicle/appUtils/ntt/ntt.cu

```diff
@@ -0,0 +1,408 @@
+                                                      // then max to allow more concurrent blocks on SM
+  const int logn_shmem = is_shared_mem_enabled ? int(log(2 * num_threads) / log(2))
+                                               : 0; // TODO: shared memory support only for types <= 32 bytes
+
+  if (inverse) {
+    if (is_shared_mem_enabled)
+      ntt_template_kernel_shared<<<num_blocks, num_threads, shared_mem, stream>>>(
+        d_inout, 1 << logn_shmem, d_twiddles, n, total_tasks, 0, logn_shmem);
+
+    for (int s = logn_shmem; s < logn; s++) // TODO: this loop also can be unrolled
+    {
+      ntt_template_kernel<E, S>
+        <<<num_blocks, num_threads, 0, stream>>>(d_inout, n, d_twiddles, n, total_tasks, s, false);
+    }
+
+    if (is_coset) batch_vector_mult(coset, d_inout, n, batch_size, stream);
+
+    num_threads = min(n / 2, MAX_NUM_THREADS);
+    num_blocks = (n * batch_size + num_threads - 1) / num_threads;
+    template_normalize_kernel<E, S>
+      <<<num_blocks, num_threads, 0, stream>>>(d_inout, n * batch_size, S::inv_log_size(logn));
+  } else {
+    if (is_coset) batch_vector_mult(coset, d_inout, n, batch_size, stream);
+
+    for (int s = logn - 1; s >= logn_shmem; s--) // TODO: this loop also can be unrolled
+    {
+      ntt_template_kernel<<<num_blocks, num_threads, 0, stream>>>(d_inout, n, d_twiddles, n, total_tasks, s, true);
+    }
+
+    if (is_shared_mem_enabled)
+      ntt_template_kernel_shared_rev<<<num_blocks, num_threads, shared_mem, stream>>>(
+        d_inout, 1 << logn_shmem, d_twiddles, n, total_tasks, 0, logn_shmem);
+  }
+
+  if (!is_sync_needed) return;
+
+  cudaStreamSynchronize(stream);
+}
+
+} // namespace
+
+template <typename S>
+void generate_twiddle_factors(S* d_twiddles, uint32_t n_twiddles, S omega, cudaStream_t stream)
+{
+  twiddle_factors_kernel<S><<<1, 1, 0, stream>>>(d_twiddles, n_twiddles, omega);
+  cudaStreamSynchronize(stream);
+}
+
+template <typename E, typename S>
+void ntt_internal(E* input, unsigned size, bool is_inverse, NTTConfig<S> config)
+{
+  uint32_t logn = uint32_t(log(size) / log(2));
+  uint32_t n_twiddles = size; // n_twiddles is set to 4096 as BLS12_381::scalar_t::omega() is of that order.
+  size_t input_size = size * config.batch_size * sizeof(E);
+  bool on_device = config.are_inputs_on_device;
+  bool generate_twiddles = (config.twiddles == nullptr);
+  
+  S* d_twiddles;
+  if (generate_twiddles) {
+    cudaMallocAsync(&d_twiddles, n_twiddles * sizeof(S), config.stream);
+    generate_twiddle_factors(d_twiddles, n_twiddles, is_inverse ? S::omega_inv(logn) : S::omega(logn), config.stream);
+  }
+
+  E* d_input;
+  if (on_device) {
+    cudaMallocAsync(&d_input, input_size, config.stream);
+    cudaMemcpyAsync(d_input, input, input_size, cudaMemcpyHostToDevice, config.stream);
+  }
+
+  if ((is_inverse && (config.ordering == Ordering::kNN || config.ordering == Ordering::kNR)) || 
```

Comment:
umm - this feels a bit clunky compared to [other libs](https://github.com/supranational/sppark/blob/fffd7349b29bb086d2a8147b227baf8b9ed006dd/ntt/ntt.cuh#L91-L110) for example